File: luogu/P1226_quick_power/main.py
#!/usr/bin/env python3
# P1226 【模板】快速幂||取余运算
# https://www.luogu.com.cn/problem/P1226
import sys


def qpow(b: int, p: int, k: int) -> int:
    r = 1 % k
    b %= k
    while p:
        if p & 1:
            r = r * b % k
        b = b * b % k
        p >>= 1
    return r


def main() -> None:
    b, p, k = map(int, sys.stdin.buffer.read().split())
    print(f"{b}^{p} mod {k}={qpow(b, p, k)}")


if __name__ == "__main__":
    main()

File: luogu/P1226_quick_power/main.rs
// P1226 【模板】快速幂||取余运算
// https://www.luogu.com.cn/problem/P1226
use std::io::{self, Read, Write};

fn qpow(mut b: i64, mut p: i64, k: i64) -> i64 {
    let mut r = 1 % k;
    b %= k;
    while p > 0 {
        if p & 1 == 1 {
            r = r * b % k;
        }
        b = b * b % k;
        p >>= 1;
    }
    r
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_ascii_whitespace();
    let b: i64 = it.next().unwrap().parse().unwrap();
    let p: i64 = it.next().unwrap().parse().unwrap();
    let k: i64 = it.next().unwrap().parse().unwrap();

    let mut out = io::BufWriter::new(io::stdout().lock());
    writeln!(out, "{}^{} mod {}={}", b, p, k, qpow(b, p, k)).unwrap();
}
